optimize/protocol.py

```diff
@@ -0,0 +1,254 @@
+        )
+
+        return stddev_estimate
+
+    return work_stddev_estimator
+
+
+def reweighted_stddev(f_n: Array, target_logpdf_n: Array, source_logpdf_n: Array) -> Float:
+    """Compute reweighted estimate of
+    stddev(f(x)) under x ~ p_target
+    based on samples   x ~ p_source
+
+    where
+        p_target(x) = exp(target_logpdf(x)) / Z_target
+
+    using samples from a different source
+        x_n ~ p_source
+        where
+        p_source(x) = exp(source_logpdf(x)) / Z_source
+
+    The inputs are arrays "{fxn_name}_n" containing the result of
+    calling each fxn on a fixed array of samples:
+
+    * f_n = [f(x_n) for x_n in samples]
+    * target_logpdf_n = [target_logpdf(x_n) for x_n in samples]
+    * source_logpdf_n = [source_logpdf(x_n) for x_n in samples]
+    """
+
+    log_weights_n = target_logpdf_n - source_logpdf_n
+    weights = np.exp(log_weights_n - logsumexp(log_weights_n)).flatten()
+
+    f_mean = np.sum(weights * f_n)
+    squared_deviations = (f_n - f_mean) ** 2
+
+    # sanitize 0 * inf -> 0 (instead of nan)
+    weighted_squared_deviations = weights * squared_deviations
+    sanitized = np.nan_to_num(weighted_squared_deviations, nan=0)
+    stddev = np.sqrt(np.sum(sanitized))
+
+    return stddev
+
+
+def construct_heuristic_lambda_pair_assessor(work_stddev_estimator) -> StepAssessor:
+    """Construct a function f(prev_lam, trial_next_lam) where bisection search on second argument
+    can be used to select next_lam so that p(x|next_lam) is a specified "distance" from p(x|prev_lam)"""
+
+    def assess_lambda_pair(prev_lam, next_lam, desired_stddev=1.0, max_step=0.25):
+        """if (next_lam - prev_lam <= max_step), compute (max(forward_stddev, reverse_stddev) - desired_stddev)"""
+        too_far = next_lam - prev_lam > max_step
+        if too_far:
+            return +np.inf
+
+        # compute max of forward, reverse work stddevs
+        forward_stddev = work_stddev_estimator(prev_lam, next_lam)
+        reverse_stddev = work_stddev_estimator(next_lam, prev_lam)
+        higher_stddev = max(forward_stddev, reverse_stddev)
+
+        return higher_stddev - desired_stddev
+
+    return assess_lambda_pair
+
+
+def greedily_optimize_protocol(assess_lambda_pair: StepAssessor, max_iterations=1000) -> Array:
+    """Optimize a lambda protocol from "left to right"
+
+    Sequentially pick next_lam so that
+    assess_lambda_pair(prev_lam, next_lam) ~= 0
+
+    assess_lambda_pair(prev_lam, next_lam) might compute
+    some measure of distance between p(x | prev_lam) and p(x | next_lam)
+    and compare that distance to a desired spacing `distance - target_distance`
+        (returning a negative number if too close, a positive number if too far)
+    """
+    protocol = [0.0]
+
+    for t in range(max_iterations):
+        prev_lam = protocol[-1]
+
+        # can we directly jump to the end?
+        if assess_lambda_pair(prev_lam, 1.0) < 0:
+            break
```

Comment:
Curious why this is necessary? Is it just an optimization, assuming that jumping to the end is more likely than any intermediate lambda?

Reply:
I think it's necessary to handle the "base case" using a clause like this, because `scipy.optimize.bisect` requires that the function changes sign in the search interval.

(The search interval is `[prev_lam, 1.0]` -- if both `assess_lambda_pair(prev_lam, prev_lam) < 0` and `assess_lambda_pair(prev_lam, 1.0) < 0`, then I think the bisection will fail.)

Reply:
got it, makes sense!